File: workspace/3_Permutaties/test/permutaties/PermutatiesTest.java
package permutaties;

import static java.util.Arrays.asList;
import static java.util.Collections.singleton;
import static java.util.Collections.singletonList;
import static junit.framework.Assert.assertEquals;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

import org.junit.Test;


public class PermutatiesTest {
	
	@Test
	public void legeSet() {
		assertEquals(new HashSet<List<Integer>>(), ListUtil.permuteer(new HashSet<Integer>()));
	}
	
	@Test
	public void setMetEenElement() {
		assertEquals(singleton(singletonList(1)), ListUtil.permuteer(asSet(1)));
	}
	
	@Test
	public void setMetTweeElementen() {
		Set<List<Integer>> expected = new HashSet<List<Integer>>();
		expected.add(asList(1, 2));
		expected.add(asList(2, 1));
		assertEquals(expected, ListUtil.permuteer(asSet(1, 2)));
	}
	
	@Test
	public void setMetDrieElementen() {
		Set<List<Integer>> expected = new HashSet<List<Integer>>();
		expected.add(asList(1, 2, 3));
		expected.add(asList(1, 3, 2));
		expected.add(asList(2, 1, 3));
		expected.add(asList(2, 3, 1));
		expected.add(asList(3, 1, 2));
		expected.add(asList(3, 2, 1));
		assertEquals(expected, ListUtil.permuteer(asSet(1, 2, 3)));
	}
	
	@Test
	public void setMetVierElementen() {
		Set<List<Integer>> expected = new HashSet<List<Integer>>();
		expected.add(asList(1, 2, 3, 4));
		expected.add(asList(1, 2, 4, 3));
		expected.add(asList(1, 3, 2, 4));
		expected.add(asList(1, 3, 4, 2));
		expected.add(asList(1, 4, 2, 3));
		expected.add(asList(1, 4, 3, 2));
		expected.add(asList(2, 1, 3, 4));
		expected.add(asList(2, 1, 4, 3));
		expected.add(asList(2, 3, 1, 4));
		expected.add(asList(2, 3, 4, 1));
		expected.add(asList(2, 4, 1, 3));
		expected.add(asList(2, 4, 3, 1));
		expected.add(asList(3, 1, 2, 4));
		expected.add(asList(3, 1, 4, 2));
		expected.add(asList(3, 2, 1, 4));
		expected.add(asList(3, 2, 4, 1));
		expected.add(asList(3, 4, 1, 2));
		expected.add(asList(3, 4, 2, 1));
		expected.add(asList(4, 1, 2, 3));
		expected.add(asList(4, 1, 3, 2));
		expected.add(asList(4, 2, 1, 3));
		expected.add(asList(4, 2, 3, 1));
		expected.add(asList(4, 3, 1, 2));
		expected.add(asList(4, 3, 2, 1));
		assertEquals(expected, ListUtil.permuteer(asSet(1, 2, 3, 4)));
	}

	private Set<Integer> asSet(Integer... integers) {
		Set<Integer> result = new HashSet<Integer>();
		for (Integer integer : integers) {
			result.add(integer);
		}
		return result;
	}

	
}

File: solution/3_Permutaties/3_Permutaties/PermutatiesTest.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Linq;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using permutaties;

namespace test
{
    [TestClass]
    public class PermutatiesTest
    {
        [TestMethod]
        public void legeSet() {
            setEquals(new HashSet<List<int>>(), ListUtil.permuteer(new HashSet<int>()));
	    }
	
	    [TestMethod]
	    public void setMetEenElement() {
            HashSet<List<int>> expected = new HashSet<List<int>>();
            expected.Add(new List<int>(new int [] {1}));

            setEquals(expected, ListUtil.permuteer(asSet(1)));
	    }
	
	    [TestMethod]
	    public void setMetTweeElementen() {
		    HashSet<List<int>> expected = new HashSet<List<int>>();
		    expected.Add(new List<int>(new int[] {1, 2}));
		    expected.Add(new List<int>(new int[] {2, 1}));
            setEquals(expected, ListUtil.permuteer(asSet(1, 2)));
	    }
	
	    [TestMethod]
	    public void setMetDrieElementen() {
		    HashSet<List<int>> expected = new HashSet<List<int>>();
		    expected.Add(new List<int>(new int[] {1, 2, 3}));
		    expected.Add(new List<int>(new int[] {1, 3, 2}));
		    expected.Add(new List<int>(new int[] {2, 1, 3}));
		    expected.Add(new List<int>(new int[] {2, 3, 1}));
		    expected.Add(new List<int>(new int[] {3, 1, 2}));
		    expected.Add(new List<int>(new int[] {3, 2, 1}));
		    setEquals(expected, ListUtil.permuteer(asSet(1, 2, 3)));
	    }
	
	    [TestMethod]
	    public void setMetVierElementen() {
		    HashSet<List<int>> expected = new HashSet<List<int>>();
		    expected.Add(new List<int>(new int[] {1, 2, 3, 4}));
		    expected.Add(new List<int>(new int[] {1, 2, 4, 3}));
		    expected.Add(new List<int>(new int[] {1, 3, 2, 4}));
		    expected.Add(new List<int>(new int[] {1, 3, 4, 2}));
		    expected.Add(new List<int>(new int[] {1, 4, 2, 3}));
		    expected.Add(new List<int>(new int[] {1, 4, 3, 2}));
		    expected.Add(new List<int>(new int[] {2, 1, 3, 4}));
		    expected.Add(new List<int>(new int[] {2, 1, 4, 3}));
		    expected.Add(new List<int>(new int[] {2, 3, 1, 4}));
		    expected.Add(new List<int>(new int[] {2, 3, 4, 1}));
		    expected.Add(new List<int>(new int[] {2, 4, 1, 3}));
		    expected.Add(new List<int>(new int[] {2, 4, 3, 1}));
		    expected.Add(new List<int>(new int[] {3, 1, 2, 4}));
		    expected.Add(new List<int>(new int[] {3, 1, 4, 2}));
		    expected.Add(new List<int>(new int[] {3, 2, 1, 4}));
		    expected.Add(new List<int>(new int[] {3, 2, 4, 1}));
		    expected.Add(new List<int>(new int[] {3, 4, 1, 2}));
		    expected.Add(new List<int>(new int[] {3, 4, 2, 1}));
		    expected.Add(new List<int>(new int[] {4, 1, 2, 3}));
		    expected.Add(new List<int>(new int[] {4, 1, 3, 2}));
		    expected.Add(new List<int>(new int[] {4, 2, 1, 3}));
		    expected.Add(new List<int>(new int[] {4, 2, 3, 1}));
		    expected.Add(new List<int>(new int[] {4, 3, 1, 2}));
            expected.Add(new List<int>(new int[] {4, 3, 2, 1}));
		    setEquals(expected, ListUtil.permuteer(asSet(1, 2, 3, 4)));
	    }

	    private HashSet<int> asSet(params int[] integers) {
		    HashSet<int> result = new HashSet<int>();
		    foreach (int integer in integers) {
			    result.Add(integer);
		    }
		    return result;
	    }

        private void setEquals(HashSet<List<int>> expected, HashSet<List<int>> actual)
        {
            Assert.AreEqual(expected.Count, actual.Count);
            foreach (List<int> list in expected)
            {
                bool inSet = false;
                foreach (List<int> actualList in actual)
                {
                    if (list.SequenceEqual(actualList))
                    {
                        inSet = true;
                    }
                }
                Assert.IsTrue(inSet);
            }
        }
    }
}
